src/status_im/chat/subs.cljs

```diff
@@ -107,80 +107,111 @@
-     :acc  (conj acc [datemark [msg]])}                      ; add new datemark group
-    {:last-timestamp (max timestamp last-timestamp)
-     :last-datemark last-datemark
-     :acc (conj (pop acc) (update (peek acc) 1 conj msg))})) ; conj to the last element
-
-(defn message-datemark-groups
-  "Transforms map of messages into sequence of `[datemark messages]` tuples, where
-  messages with particular datemark are sorted according to their clock-values."
+  (cond (empty? acc)                                     ; initial element
+        {:last-timestamp timestamp
+         :last-datemark  datemark
+         :acc            (conj acc msg)}
+
+        (and (not= last-datemark datemark)               ; not the same day
+             (< timestamp last-timestamp))               ; not out-of-order
+        {:last-timestamp timestamp
+         :last-datemark  datemark
+         :acc            (conj acc {:value last-datemark ; intersperse datemark message
+                                    :type  :datemark}
+                               msg)}
+        :else
+        {:last-timestamp (max timestamp last-timestamp)  ; use last datemark
+         :last-datemark  last-datemark
+         :acc            (conj acc (assoc msg :datemark last-datemark))}))
+
+(defn sort-messages
+  "Remove hidden messages and sort by clock-value desc, breaking ties by message id"
   [id->messages]
-  (let [sorted-messages     (->> id->messages
-                                 vals
-                                 (sort-by (juxt (comp unchecked-negate :clock-value) :message-id))) ; sort-by clock in reverse order, break ties by :message-id field
-        remove-hidden-xf    (filter :show?)
-        add-datemark-xf     (map (fn [{:keys [timestamp] :as msg}]
-                                   (assoc msg :datemark (time/day-relative timestamp))))]
-    (-> (transduce (comp remove-hidden-xf
-                         add-datemark-xf)
-                   (completing partition-by-datemark)
-                   {:acc []}
-                   sorted-messages)
-        :acc)))
-
-(reg-sub
-  :get-chat-message-datemark-groups
+  (->> id->messages
+       vals
+      (filter :show?)
+      (sort-by (juxt (comp unchecked-negate :clock-value) :message-id))))
+
+(defn- add-datemark [{:keys [timestamp] :as msg}]
+   (assoc msg :datemark (time/day-relative timestamp)))
+
+(defn- add-timestamp [{:keys [timestamp] :as msg}]
+  (assoc msg :timestamp-str (time/timestamp->time timestamp)))
+
+(defn intersperse-datemarks
+  "Add a datemark in between an ordered seq of messages when two datemarks are not
+  the same. Ignore messages with out-of-order timestamps"
+  [messages]
+  (when (seq messages)
+    (let [messages-with-datemarks (->> (transduce (comp
+                                                    (map add-datemark)
+                                                    (map add-timestamp))
+                                                  (completing intersperse-datemark)
```

Comment:
We could just pass `:acc` as completing function here. Then no need to use threading macro.